### ISSUE TITLE
OIDC init failure, relative URL without a base

### ISSUE THREAD

**Issue by joehand** [mbecker20/komodo#242]
I'm using [Pocket Id](https://github.com/stonith404/pocket-id) for OIDC auth. Its working with other services fine. 

I've tried a few configs to get it going with Komodo but unable to get it working. As far as I can tell, the OIDC_PROVIDER url I should use is the root (and that is what I use on other setups). Its on a subdomain via Caddy reverse proxy, for example `https://id.0jjj.casa` (not the actual domain but similar and it does use the casa TLD, not sure if that is causing the issue). I've confirmed that `https://id.0jjj.casa/.well-known/openid-configuration` works fine and is accessible from Komodo.

```
oidc_provider = "https://id.0jjj.casa" 

thread 'tokio-runtime-worker' panicked at bin/core/src/auth/oidc/client.rs:66:4:
called `Result::unwrap()` on an `Err` value: Failed to init default OIDC client

Caused by:
    relative URL without a base
```

I've tried other URLs, which can connect. But then there is a mismatch with the issuer, such as the via docker network:

```
oidc_provider = "http://pocket-id:3000"

Caused by:
    0: Failed to get OIDC /.well-known/openid-configuration
    1: Validation error: unexpected issuer URI `https://id.0jjj.casa` (expected `http://pocket-id:3000`)
```

**Comment by mbecker20** [comment]
Are you running your OIDC provider on the same host as Komodo?

**Comment by mbecker20** [comment]
How is it configured for other services? Eg what is used as their provider URLS, and is there any relevant networking differences. The configuration here should follow what you are using there. For example, I don't see any trailing application name `https://id.0jjj.casa/komodo/` which is often (but not always) part of OIDC.

For supported feature reference, the code handling OIDC in Komodo is here: https://github.com/ramosbugs/openidconnect-rs

**Comment by joehand** [comment]
I wrote a lot below. Not sure its relevant. Turns out I think its the domain causing issues since it starts with a number.

If I change `https://id.0jjj.casa` to `https://id.jjj.casa` it fails with a different error - failed to get OIDC config, as expected (since the domain doesn't exist), instead of the relative URL without a base. So I believe the number at the start of the domain is causing the issue.

However, it seems to pass a test fine in upstream libraries, so not sure exactly what is going on. It doesn't seem to be an upstream issue (assuming I understand how these tests work...). I added [two tests for it](https://github.com/joehand/openidconnect-rs/pull/1/files) on openidconnect-rs and they seem to pass.

Would [this be relevant](https://github.com/ramosbugs/openidconnect-rs/issues/77#issuecomment-1738092791)? I see Komodo is slightly different in [how it creates the IssuerUrl](https://github.com/mbecker20/komodo/blob/e274d6f7c8a3ab6f6ef53d3ea0b547ac57e9801b/bin/core/src/auth/oidc/client.rs#L37).

---

Thanks for linking the OIDC library you use, let me know if you want me to report upstream there. It looks like [trailing slashes are automatically added](https://github.com/ramosbugs/openidconnect-rs/issues/77) to issuer url if missing before discovery to confirm to [the spec](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfig) (which is discussed in the repo you linked):

> OpenID Providers supporting Discovery MUST make a JSON document available at the path formed by concatenating the string `/.well-known/openid-configuration` to the Issuer.

> How is it configured for other services? 

Generally always just with the root domain, specific services don't need/get applications names in the provider URLs. I did try to proxy `auth.domain/komodo/*` to `auth.domain` and it fixed the relative URL error but then I got the issuer URL error instead which I can't seem to get around.

Here is a working [Outline config](https://docs.getoutline.com/s/hosting/doc/oidc-8CPBm6uC0I): 

```
OIDC_CLIENT_ID=[[OUTLINE_OIDC_CLIENT_ID]]
OIDC_CLIENT_SECRET=[[OUTLINE_OIDC_CLIENT_SECRET]]
OIDC_AUTH_URI=https://id.0jjj.casa/authorize
OIDC_TOKEN_URI=https://id.0jjj.casa/api/oidc/token
OIDC_USERINFO_URI=https://id.0jjj.casa/api/oidc/userinfo
```

I'm using [OAuth-2proxy](https://oauth2-proxy.github.io/oauth2-proxy/) in a few places and this is the config:

```
# (Client ID/Secret set elsewhere)
oidc_issuer_url="https://id.0jjj.casa"
provider="oidc"
provider_display_name="Pocket Id"
scope = "openid email"
```

I had Zitadel setup before moving to PocketId. I just tried to use that and am facing the same issues. It similarly does not have an application-specific path for each discovery endpoint. (Here is [their docs on NextCloud setup](https://zitadel.com/blog/zitadel-as-sso-provider-for-selfhosting), if relevant).


<details>
<summary>Full backtrace on *relative URL without a base*</summary>

```
thread 'tokio-runtime-worker' panicked at bin/core/src/auth/oidc/client.rs:66:4:
called `Result::unwrap()` on an `Err` value: Failed to init default OIDC client

Caused by:
    relative URL without a base

Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
   1: <tokio::future::maybe_done::MaybeDone<Fut> as core::future::future::Future>::poll
   2: <core::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
   3: core::app::{{closure}}
   4: tokio::runtime::task::core::Core<T,S>::poll
   5: tokio::runtime::task::harness::Harness<T,S>::poll
   6: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
   7: tokio::runtime::scheduler::multi_thread::worker::Context::run
   8: tokio::runtime::context::runtime::enter_runtime
   9: tokio::runtime::scheduler::multi_thread::worker::run
  10: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
  11: tokio::runtime::task::core::Core<T,S>::poll
  12: tokio::runtime::task::harness::Harness<T,S>::poll
  13: tokio::runtime::blocking::pool::Inner::run
  14: std::sys::backtrace::__rust_begin_short_backtrace
  15: core::ops::function::FnOnce::call_once{{vtable.shim}}
  16: std::sys::pal::unix::thread::Thread::new::thread_start
  17: start_thread
  18: clone
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: <tokio::future::maybe_done::MaybeDone<Fut> as core::future::future::Future>::poll
   4: <core::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
   5: core::app::{{closure}}
   6: tokio::runtime::task::core::Core<T,S>::poll
   7: tokio::runtime::task::harness::Harness<T,S>::poll
   8: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
   9: tokio::runtime::scheduler::multi_thread::worker::Context::run
  10: tokio::runtime::context::runtime::enter_runtime
  11: tokio::runtime::scheduler::multi_thread::worker::run
  12: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
  13: tokio::runtime::task::core::Core<T,S>::poll
  14: tokio::runtime::task::harness::Harness<T,S>::poll
  15: tokio::runtime::blocking::pool::Inner::run
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Error: task 5 panicked with message "called `Result::unwrap()` on an `Err` value: Failed to init default OIDC client\n\nCaused by:\n    relative URL without a base\n\nStack backtrace:\n   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from\n   1: <tokio::future::maybe_done::MaybeDone<Fut> as core::future::future::Future>::poll\n   2: <core::future::poll_fn::PollFn<F> as core::future::future::Future>::poll\n   3: core::app::{{closure}}\n   4: tokio::runtime::task::core::Core<T,S>::poll\n   5: tokio::runtime::task::harness::Harness<T,S>::poll\n   6: tokio::runtime::scheduler::multi_thread::worker::Context::run_task\n   7: tokio::runtime::scheduler::multi_thread::worker::Context::run\n   8: tokio::runtime::context::runtime::enter_runtime\n   9: tokio::runtime::scheduler::multi_thread::worker::run\n  10: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll\n  11: tokio::runtime::task::core::Core<T,S>::poll\n  12: tokio::runtime::task::harness::Harness<T,S>::poll\n  13: tokio::runtime::blocking::pool::Inner::run\n  14: std::sys::backtrace::__rust_begin_short_backtrace\n  15: core::ops::function::FnOnce::call_once{{vtable.shim}}\n  16: std::sys::pal::unix::thread::Thread::new::thread_start\n  17: start_thread\n  18: clone"

Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
   1: core::main::{{closure}}
   2: tokio::runtime::park::CachedParkThread::block_on
   3: tokio::runtime::context::runtime::enter_runtime
   4: core::main
   5: std::sys::backtrace::__rust_begin_short_backtrace
   6: std::rt::lang_start::{{closure}}
   7: std::rt::lang_start_internal
   8: main
   9: __libc_start_main
  10: _start
```
</details>

**Comment by mbecker20** [comment]
I have setup Zitadel, and you need to use this part: https://github.com/mbecker20/komodo/blob/main/config/core.config.toml#L183

In terms of this one, did you try with trailing `/`? `oidc_provider = "https://id.0jjj.casa/"`

**Comment by joehand** [comment]
Success! Sorry about the hassle, user error. `KOMODO_HOST` wasn't set causing the redirect URI creation to fail.

A few learnings:

1. I had removed `KOMODO_HOST` from the config by accident and it was never set. I know it says "required" but never got any errors without it being set, so may want to enforce that or throw something in logs if its not set.
2.  My local username was `joe`. On first try logging in with OIDC, I got an error because it has same username and `KOMODO_OIDC_USE_FULL_EMAIL` was false. I updated local username and was able to create new user with no issue.

```
'Kind: An error occurred when trying to execute a write operation: WriteError(WriteError { code: 11000, code_name: None, message: "E11000 duplicate key error collection: komodo.User", details: None }), labels: {}'
```

It's really been fun to get Komodo setup and running, thanks for the great work! I may try to contribute a bit to the docs because there were some pieces scattered across various places that required a bit of digging.
